storage/innobase/trx/trx0trx.cc

```diff
@@ -385,6 +385,7 @@ void trx_t::free()
   check_foreigns= true;
   assert_freed();
   trx_sys.rw_trx_hash.put_pins(this);
+  need_wait_report= false;
```

Comment:
It would be cleaner to clear the `need_wait_report` in `trx_t::commit_cleanup()` or similar, and extend `assert_freed()` with `ut_ad(!need_wait_report)`.

Reply:
@dr-m Hm, I don't understand. `need_wait_report` is only set once in the thread, in `innobase_trx_allocate()` when the trx_t object is first allocated to the thread. But trx_t::commit_cleanup() is called after every commit, right?

The value of `need_wait_report` needs to be preserved until the trx_t object is no longer associated with the thread. Is there a more appropriate place for this than `trx_t::free()` ?

Reply:
There is a special `trx_pools` that manages transaction objects, so that some memory allocation overhead and fragmentation will be avoided. Most members of `trx_t` are reset on transaction commit, in `trx_t::commit_cleanup()`, and validated in `trx_t::assert_freed()`. If you added the new data member next to `unsigned check_unique_secondary:1`, then I believe that resetting the field in `trx_t::commit_cleanup()` would incur no overhead at all; several bits would be cleared in the same instruction.